shell/renderer/api/atom_api_context_bridge.cc

```diff
@@ -0,0 +1,408 @@
+                     RenderFramePersistenceStore* store,
+                     size_t func_id) {
+    new FunctionLifeMonitor(isolate, target, store, func_id);
+  }
+
+ protected:
+  FunctionLifeMonitor(v8::Isolate* isolate,
+                      v8::Local<v8::Object> target,
+                      RenderFramePersistenceStore* store,
+                      size_t func_id)
+      : ObjectLifeMonitor(isolate, target), store_(store), func_id_(func_id) {}
+  ~FunctionLifeMonitor() override = default;
+
+  void RunDestructor() override { store_->functions().erase(func_id_); }
+
+ private:
+  RenderFramePersistenceStore* store_;
+  size_t func_id_;
+};
+
+}  // namespace
+
+RenderFramePersistenceStore::RenderFramePersistenceStore(
+    content::RenderFrame* render_frame)
+    : content::RenderFrameObserver(render_frame) {}
+
+RenderFramePersistenceStore::~RenderFramePersistenceStore() = default;
+
+void RenderFramePersistenceStore::OnDestruct() {
+  delete this;
+}
+
+v8::Local<v8::Value> PassValueToOtherContext(
+    v8::Local<v8::Context> source,
+    v8::Local<v8::Context> destination,
+    v8::Local<v8::Value> value,
+    RenderFramePersistenceStore* store) {
+  // Proxy functions and monitor the lifetime in the new context to release
+  // the global handle at the right time.
+  if (value->IsFunction()) {
+    auto func = v8::Local<v8::Function>::Cast(value);
+    v8::Global<v8::Function> global_func(source->GetIsolate(), func);
+    v8::Global<v8::Context> global_source(source->GetIsolate(), source);
+
+    size_t func_id = store->take_id();
+    store->functions()[func_id] =
+        std::make_tuple(std::move(global_func), std::move(global_source));
+    {
+      v8::Context::Scope destination_scope(destination);
+      v8::Local<v8::Value> proxy_func = mate::ConvertToV8(
+          destination->GetIsolate(),
+          base::BindRepeating(&ProxyFunctionWrapper, store, func_id));
+      FunctionLifeMonitor::BindTo(destination->GetIsolate(),
+                                  v8::Local<v8::Object>::Cast(proxy_func),
+                                  store, func_id);
+      return proxy_func;
+    }
+  }
+
+  // Proxy promises as they have a safe and guarunteed memory lifecycle (unlike
+  // functions)
+  if (value->IsPromise()) {
+    v8::Context::Scope destination_scope(destination);
+
+    auto v8_promise = v8::Local<v8::Promise>::Cast(value);
+    auto* promise =
+        new util::Promise<v8::Local<v8::Value>>(destination->GetIsolate());
+    v8::Local<v8::Promise> handle = promise->GetHandle();
+
+    auto then_cb = base::BindOnce(
+        [](util::Promise<v8::Local<v8::Value>>* promise, v8::Isolate* isolate,
+           v8::Global<v8::Context> source, v8::Global<v8::Context> destination,
+           RenderFramePersistenceStore* store, v8::Local<v8::Value> result) {
+          promise->Resolve(PassValueToOtherContext(
+              source.Get(isolate), destination.Get(isolate), result, store));
+          delete promise;
```

Comment:
What happens if the RenderFrame is destroyed before the promise resolves? Does this leak?

Reply:
My understanding is that the context is destroyed before the frame is destroyed, when the context is destroyed all V8 objects are GC'ed so this callback that is bound to V8 will be GC'ed which will then release everything else.